### DISCUSSION TITLE
Error messages from USER_FUNC2 setting?

### DISCUSSION THREAD

**Discussion by ArjenR49** [PiSupply/PiJuice#960] [Q&A] [answered]
I have been trying out different ideas/strategies for shutting down a Pi4B which has a PiJuice HAT with the standard battery. Power is supplied via the Pi's USB type C connector, as that is the power supply for a Pi4B which I have available.

The script below is assigned to be executed as a user function when the No Power event occurs.
I've used another script as a user function and that worked.

The code is as follows:

```
#!/usr/bin/python3

# See: https://learn.pi-supply.com/make/how-to-run-a-user-script-with-the-pijuice-power-events/

import os
import logging
from time import sleep
from pijuice import PiJuice

logging.basicConfig(
    filename = '/home/pi/PiShutdown.log',
    level = logging.DEBUG,
    format = '%(asctime)s %(message)s',
    datefmt = '%d/%m/%Y %H:%M:%S')

# Write statement to log
logging.info('Running on PiJuice/UPS battery ... shutdown imminent ...')

pijuice = PiJuice(1, 0x14)

# Remove power to PiJuice MCU IO pins
pijuice.power.SetSystemPowerSwitch(0)

# Set wakeup
pijuice.power.SetWakeUpOnCharge(0.0)

# Remove 5V power to RPi after 60 seconds
pijuice.power.SetPowerOff(60)

# Show custom script is run by blinking the user LED red 10x
pijuice.status.SetLedBlink('D2', 10, [200,0,0], 50, [0, 0, 0], 50)
sleep(2)

# Now shut down
os.system("sudo shutdown -h now")


```

When I run this file manually, the LED blinks and the Pi is shut down as one would expect.
However, upon removal of AC power the No Power event either doesn't occur, or something else is wrong.
One way or another on my last try I got the following error messages in the log file set up by the script: 

```
21/03/2023 00:44:15 pickle loaded: /usr/lib/python3/dist-packages/jedi/third_party/typeshed/stdlib/2and3/builtins.pyi
21/03/2023 00:44:16 diff parser start
21/03/2023 00:44:16 line_lengths old: 37; new: 37
21/03/2023 00:44:16 -> code[equal] old[1:30] new[1:30]
21/03/2023 00:44:16 copy old[5:28] new[1:28]
21/03/2023 00:44:17 parse_part from 29 to 31 (to 31 in part parser)
21/03/2023 00:44:17 -> code[replace] old[31:31] new[31:31]
21/03/2023 00:44:17 -> code[equal] old[32:37] new[32:37]
21/03/2023 00:44:17 copy old[32:35] new[32:35]
21/03/2023 00:44:17 parse_part from 36 to 37 (to 36 in part parser)
21/03/2023 00:44:17 diff parser end
21/03/2023 00:44:19 diff parser start
21/03/2023 00:44:19 line_lengths old: 37; new: 37
21/03/2023 00:44:19 -> code[equal] old[1:30] new[1:30]
21/03/2023 00:44:19 copy old[5:28] new[1:28]
21/03/2023 00:44:19 parse_part from 29 to 31 (to 31 in part parser)
21/03/2023 00:44:19 -> code[replace] old[31:31] new[31:31]
21/03/2023 00:44:19 -> code[equal] old[32:37] new[32:37]
21/03/2023 00:44:19 copy old[32:35] new[32:35]
21/03/2023 00:44:19 parse_part from 36 to 37 (to 36 in part parser)
21/03/2023 00:44:19 diff parser end
21/03/2023 00:44:19 pickle loaded: /usr/lib/python3/dist-packages/pijuice.py
21/03/2023 00:44:33 Running on PiJuice/UPS battery ... shutdown imminent ...

```

The last line is the only intended output and what I get when I run the script manually.
As the No Power event with this script doesn't currently start the assigned script, I'm glad that some error messages were logged, but I have no clue what I could do to prevent them. The script is executable. 
As I wrote, I ran the script manually (in Thonny and from a terminal) and it did exactly what is intended.

Using the No Power event with a user function script looks like the most straightforward way to make the Pi shut down on loss of AC Power. No delay timing loops on the Pi and a simple script as a user function.
 
The Pi4 has an SSD attached and according to the PiJuice GUI the currents through the GPIO pins involved are typically over 1 A and close to 2 A and peaking at 2,5 A, which is comparatively high for the standard battery and the PiJuice.
Therefore my latest plan is to make the Pi shut down without undue delay on loss of AC power and have it boot automatically when power is restored, but it doesn't quite work, yet ...

**Comment by ArjenR49** [comment, marked as answer]
Enabling System Events solved the problem, although the error messages remain mysterious to me.
None of the listed system events needs to be enabled, just the upper box ...

Maybe in the past - I got this PiJuice HAT already in 2019, but haven't used it after initial testing - you couldn't get the Pi to start upon return of AC power in this configuration where the power supply is connected to the Pi (not much choice, Pi4 has a different connector), but it has worked consistently with the current software 1.8 and firmware 1.6. No need to press any buttons. 
Based on my experiences today and yesterday, I find the PiJuice HAT the most reliable UPS I know. I have had a few other UPS's. Not the boards with batteries that aren't UPS's, but made by Omzlo, PiModules, Olmatic and 52Pi/GeeekPi. It's just that I haven't found a good source of documentation. Instead I found and copied PiJuice specific commands from code written by others.
Searching for a manual will find one from a dozen sources the same nice looking 12 page brochure containing next to no information. Still, the pieces of documentation I find here and there is more than for the other four UPS's I mentioned and this forum seems alive, too, which is great!

**Comment by tvoverbeek** [comment]
You are not enabling the events, but the system task.
The system task running on the RPi monitors the power inputs and executes the user functions + system functions. For the nitty gritty look at the pijuice_sys.py source code (https://github.com/PiSupply/PiJuice/blob/master/Software/Source/src/pijuice_sys.py).
Note: you can achieve the same (minus the logging) by defining SYS_FUNC_HALT_POW_OFF for the no power event.

You can add logging for all shutdowns by using a user function for the system stop event. This event gets triggered when the system task (pijuice service) is terminated as part of the system shutdown => Do not call shutdown from this user function since the system shutdown is already in progress.

The initial messages have nothing to do with the PiJuice itself. They are from the Jedi software which must be running on your system.
It seems to be checking there are no changes to the main pijuice module.

**Reply by ArjenR49** [reply]
I tried what you suggested, using SYS_FUNC_HALT_POW_OFF for the no power event. Disconnecting AC power to the Pi the PiJuice did stop the Pi, as expected, but I wasn't then able to start the Pi by reconnecting AC.

I also tried adding logging the system stop event as you suggested. Code to flash a LED that I added to the user function for the system stop event was executed, which reveals something about when shutdown caused by the no power event actually occurs.
 
However, the show stopper for using the provided function was that I had to press SW1 to start the Pi and then having to do it while the power is still off. With AC Power reconnected I cannot restart the Pi. Perhaps it would work if I connect the power supply to the PiJuice board.  

Using the No Power event script that I posted above I don't have that problem. The Pi starts up right away when I reconnect AC power. I tend to think perhaps it is the PiJuice setting commands in the script that cause the difference.

The code in pijuice_sys.py you gave as study material is quite a bit over my head and it looks like calling functions outside of the code file at hand. (Object Oriented Programming? There was only one fleeting moment in a SAS course long ago when I thought I might understand OOP :-(

A particular command in pijuice_sys.py apparently 'translates' into two different functions and another command into one function and so on, but based on that code alone, I am not able to tell why the PiJuice setting commands in the No Power event script I posted above are required and can only guess at their effect. There must be a manual of sorts with that kind of info, as the writer of the article that I adapted the script from must have had a source.
I have been looking for a link to a manual in the fancy little square 12 page brochure from the moment I opened the PiJuice box, but probably it was too early days back then.

**Comment by tvoverbeek** [comment]
@ArjenR49 If you shut down with SYS_FUNC_HALT_POW_OFF due to no power disconnected from the PI, the PiJuice continues to supply 5V to the Pi for 30 sec. PiJuice LED L1 is still on (green if battery > 50%). After the 30 sec (60 sec in your USER_FUNC2) L1 goes off and changes to a very weak slow blink. If you reconnect the power to the Pi before the 30 sec has expired the Pi will not boot (since it never was without 5V) and you get the symptoms you observed.
When there is no 5V for the Pi, reconnecting power should boot the Pi independent from the PiJuice being mounted or not.

Anyway since your user script does what you want keep using it.

**Reply by ArjenR49** [reply]
When the power to the Pi (USB ps) returns within the set time (30/60 seconds) before the PiJuice has cut power to the Pi from the battery, too, the Pi has halted, but will not start up ever again w/o an operator intervening somehow ...
Is that the right conclusion?
Do you know if this can be prevented and if so, how?

There is this elusive event called Power present. Should I look into that (I've not found anything about it, yet)?

**Comment by tvoverbeek** [comment]
Yes, that is the right conclusion.
When the Pi has halted (shutdown completed) it needs a 5v power interruption in order to start.

If you know how long your shutdown process takes, set the timeout to just a bit longer.
Then you minimize the window where returning power cancels the pijuice switching off the 5V to the GPIO.
Power present event detection depends on the pijuice service running -> will not happen when the Pi is off/halted.
It can be used to detect returning external power when running on battery

**Reply by ArjenR49** [reply]
I experimented with the return of AC power in various phases while the loss of AC has already triggered my shutdown script to run.

0. AC power removed, but PiJuice has not detected the new situation yet.
1. As soon as PiJuice detects the loss of AC, the script starts running, flashes D2 and writes several lines to the log
2. Next it waits for 30 seconds, flashes D2 again and the issues the shutdown -h command
3. The Pi is shutting down (takes some 30 seconds in my case)
4. Pi has halted (green led flashed 10 times)
5. PiJuice supplies battery power to the Pi for some more time (for 45 seconds after shutdown -h command)
6. Battery power cut, Pi's red led off.

In subsequent experiments I reconnected AC to the Pi during anyone of these phases.
The script (see below) does keep on running - as seen from visible signs like LED flashes on PiJuice and Pi and the log entries, but **eventually the Pi always starts up again**.

This is very welcome, but it contradicts what I was thinking, viz. Pi will not start up if AC power is returned during phase 5.  
In phase 5 the Pi still has battery power (red led on), but has stopped running. Still the Pi restarts when I return AC power at that point. 
The Power present event doesn't happen, as the Pi is not running, like [you](https://github.com/tvoverbeek) said.

One might expect the Power present event to happen when quickly reconnecting AC power in phase 0 (Pi running on battery), but I didn't see it. Probably just because the PiJuice hasn't yet noticed.
I will have to set up a different experiment, if I want to find out when Power present actually will happen. ( ... disable No power event so the Pi isn't shut down by the assigned script, but much later when the battery charge falls below the threshold ... and return AC before that ...)
EDIT: Power present event is indeed triggered as described in the experiment as planned above. It takes a little while before it happens and the assigned script writes to the log. Doing something more useful based on this event is another thing ...

The goal of this 'No power' script can, of course, be achieved with a lot less code or perhaps even without, but I started this over a year ago when I was much more at a loss about how to get PiJuice to do what I want (make sure the Pi's file system is OK after a blackout).

As for setting up for a restart upon return of AC, my script only has these two lines:
```
pijuice.rtcAlarm.SetWakeupEnabled(True)
pijuice.power.SetWakeUpOnCharge(50, non_volatile=True)

```
Wakeup on charge % is at 80, minimum charge 50, minimum battery voltage 3,4.

The full script:
```
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# See: https://learn.pi-supply.com/make/how-to-run-a-user-script-with-the-pijuice-power-events/
# ar - 24-03-2023, 15-04-2023, 20-04-2023, 18-05-2023, 13-06-2023, 14-06-2023, 20.11.2023, 24-02-2024,
#      24-05-2024, 25-05-2024, 27-05-2024, 31-05-2024

import os
import logging
from time import sleep
from pijuice import PiJuice

logging.basicConfig(
    filename = '/home/pi/PiJuiceUPS/PiJuiceStatus.log',
    level = logging.DEBUG,
    format = '%(asctime)s %(message)s',
    datefmt = '%d-%m-%Y %H:%M:%S')

pijuice = PiJuice(1, 0x14)
# Time in seconds to keep running on battery before issuing halt command.
# Must be longer than the first blinking (10 seconds) 
HaltDelay = 30
# Delay for PiJuice to cut battery power after halt command has been issued.
PowerCutDelay = 45

# Make sure PiJuice status is OK.
pjOK = False
while pjOK == False:
    stat = pijuice.status.GetStatus()
    if stat['error'] == 'NO_ERROR':
        pjOK = True
    else:
        sleep(0.1)

# Write message to log file
logging.info('AC power loss -> initiating halt in ' + str(HaltDelay) + ' seconds.')
logging.info('Pi will auto-restart upon return of AC power.')

# Remove power to PiJuice VSYS output (for peripherals)
pijuice.power.SetSystemPowerSwitch(0)

# Make sure wakeup enabled and wakeup_on_charge have the correct values
pijuice.rtcAlarm.SetWakeupEnabled(True)
pijuice.power.SetWakeUpOnCharge(50, non_volatile=True)

# Show custom script runs by blinking user LED in orange 10x
logging.info('Blink LED and then wait ' + str(HaltDelay) + ' seconds until halt command.')
pijuice.status.SetLedBlink('D2', 10, [255,160,0], 500, [0,0,0], 500)
sleep(HaltDelay)

# PiJuice cuts 5V power to RPi PowerCutDelay seconds after halt command.
pijuice.power.SetPowerOff(PowerCutDelay)

# Now blink LEDs again & issue halt command
logging.info('Halt imminent ... battery power will be cut ' + str(PowerCutDelay) + ' seconds from now.')
pijuice.status.SetLedBlink('D2', 5, [100,0,0], 100, [0,0,0], 50)
sleep(0.75)
# Issue halt command
os.system("sudo shutdown -h now")

```
